File: Problem solving with c++/Problem solving Level 2/Problem06.cpp
/*
Problem06 |===============================================

Write a program to read a number and print all the sum of
its digits

Input : 
1234

Output :

sum of Digit = 10

==========================================================
*/


#include <iostream>
using namespace std;

/**
 * Function to read a positive integer from the user.
 * It ensures that the number entered is non-negative.
 *
 * @param message - The prompt message displayed to the user.
 * @return The validated positive integer.
 */
int ReadNumberPositive(string message) {
    int Number;
    do {
        cout << message;
        cin >> Number;
    } while (Number < 0); // Loop until a non-negative number is entered
    return Number;
}

/**
 * Function to calculate and print the sum of the digits of a given number.
 *
 * @param Number - The number whose digits will be summed.
 */
void CheckSumRemainder(int Number) {
    int Remainder;
    int SumRemainder = 0;
    do {
        Remainder = Number % 10; // Extract the last digit
        SumRemainder += Remainder; // Add it to the sum
        Number = Number / 10; // Remove the last digit
    } while (Number > 0);

    cout << "Sum of digits: " << SumRemainder << endl;
}

/**
 * Main function - Entry point of the program.
 * It asks the user for input and calculates the sum of its digits.
 */
int main() {
    // Read a positive number and pass it to CheckSumRemainder function
    CheckSumRemainder(ReadNumberPositive("Enter a positive number to calculate the sum of its digits: "));

    return 0;
}

File: Problem solving with Python/Problem06.py
"""
Problem06 |===============================================

Write a program to read a number and print all the sum of
its digits

Input : 
1234

Output :

sum of Digit = 10

==========================================================
"""


def read_positive_number(message):
    """
    Function to read a positive integer from the user.
    Ensures that the input is non-negative.
    
    :param message: The prompt message displayed to the user.
    :return: The validated positive integer.
    """
    number = -1  # Initialize with an invalid value
    while number < 0:
        try:
            number = int(input(message))  # Read and convert input to an integer
            if number < 0:
                print("Please enter a non-negative integer.")
        except ValueError:
            print("Invalid input. Please enter an integer.")
    return number

def calculate_sum_of_digits(number):
    """
    Function to calculate and print the sum of the digits of a given number.

    :param number: The number whose digits will be summed.
    """
    sum_digits = 0  # Variable to store the sum of digits
    
    while number > 0:
        remainder = number % 10  # Extract the last digit
        sum_digits += remainder  # Add it to the sum
        number //= 10  # Remove the last digit
    
    print(f"Sum of digits: {sum_digits}")

# Main entry point
if __name__ == "__main__":
    num = read_positive_number("Enter a positive number to calculate the sum of its digits: ")
    calculate_sum_of_digits(num)

